crates/admin/src/cluster_controller/service.rs

```diff
@@ -243,114 +243,148 @@ impl<T: TransportConnect> Service<T> {
         }
     }
 
-    pub async fn run(mut self) -> anyhow::Result<()> {
-        self.init_partition_table().await?;
+    async fn next_cluster_state(
+        &self,
+        state: &mut ClusterControllerState<T>,
+    ) -> anyhow::Result<()> {
+        let nodes_config = self.metadata.nodes_config_ref();
+        let maybe_active = nodes_config
+            .get_admin_nodes()
+            .filter(|node| {
+                self.observed_cluster_state
+                    .is_node_alive(node.current_generation)
+            })
+            .map(|node| node.current_generation.as_plain())
+            .sorted()
+            .next();
 
-        let bifrost_admin = BifrostAdmin::new(
-            &self.bifrost,
-            &self.metadata_writer,
-            &self.metadata_store_client,
-        );
+        // A Cluster Controller is active if the node holds the smallest PlainNodeID
+        // If no other node was found to take leadership, we assume leadership
 
-        let mut shutdown = std::pin::pin!(cancellation_watcher());
-        let mut config_watcher = Configuration::watcher();
-        let mut cluster_state_watcher = self.cluster_state_refresher.cluster_state_watcher();
+        let is_leader = match maybe_active {
+            None => true,
+            Some(leader) => leader == self.metadata.my_node_id().as_plain(),
```

Comment:
GenerationalNodeId has derive Ord implementation. It means Ids with smaller generation id will show up first after sorting. 

I did it then like this in case there is the same node is somehow still alive with 2 different generation. But now I am wondering if this can actually happen.